### PR TITLE
Autoformer search space

### PR DESCRIPTION
**Pull request by nbl97**
### Description ###
We build the search space of Autoformer in `nni/examples/nas/hub/searchspace/autoformer.py`.

There are 4 searchable variables, depth, embedding dimension, heads number and MLP ratio. 
-	Depth, the number of layers of the architecture.
-	Embedding dimension, the embedding dimension in each layer. Note that it is shared in each layer which means the architecture is non-hierarchical. 
-	Heads number, the number of heads in self-attn in each layer. The heads number in each layer is independent, like the kernel size in traditional CNN search space.
-	MLP ratio, the ratio in FFN in each layer. Like the heads number, the MLP ratio is also independent in each layer.

For depth and embedding dimension, we use nn.ValueChoice() to select the values and pass the embedding dimension as a parameter to each layer to keep the dimension consistent.
For the heads number and MLP ratio, we use nn.LayerChoice() to select from the candidates in each layer.

The weight entanglement technique is not supported in the current version. We will follow the latest NNI framework and update the search space in time.

### Checklist ###
  - [x] test case
  - [x] doc

### How to test ###
We used the coming features of nn.ValueChoice(), such as multiplying the value with a constant or passing the value to nn.Parameter(). We will test the search space once the new features are supported. I look forward to more discussion about Autoformer_Space2NNI.

### CODE REVIEW

**Comment by ultmaster** [comment]
Hi there,

I briefly go through the code, the only thing that is troublesome here is that you are using ValueChoice in ``nn.Parameter(torch.zeros(xxx, xxx))``. I haven't given much thought about it, but at first glance, that might be a little bit tricky to support.

The other parts look just fine.

A few minor advices / questions here:

* ValueChoice can NOT be used in a for-loop. You need to use the ``nn.Repeat`` API to represent a mutable depth.
* ``set_sample_config`` is not needed. NNI is born for that.
* I notice ``calc_dropout``. Do you have a dynamic dropout based on the sampled ``embed_dim``?
* What's the difference between ``AttentionSuper`` and PyTorch's official ``MultiHeadAttention``? Since we are working on an official super-kernel support for PyTorch's built-in operators (including MultiHeadAttention), it would be better to use the PyTorch's official implementation, so that we can leverage each other's efforts to implement weight entanglement later.

**Comment by nbl97** [comment]
Hi @ultmaster ,
We are extremely sorry for the late reply. Thanks for your careful review and valuable suggestions. The following are our responses

- The dimension of absolute positional embedding and cls_token are both associated to the sampled embedding dimension, so I used ValueChoice in `nn.Parameter` to generate they two. We can discuss how to avoid the trouble.
  
- I have used the `nn.Repeat` API for ValueChoice, instead of using ValueChoice in a for-loop.
  
- The `set_sample_config` has been moved. Sorry for the uncleaned code.
  
- The dynamic `calc_dropout` has been moved. Since drop_out (NOT drop_path_rate) and attn_drop_out have been set to 0 in most vision transformer works up to now, dynamic `calc_dropout` has not been fully tested, so we decide to remove this feature first.
  
- The `AttentionSuper` was designed to support optional relative positional embedding and some tricky implementation of the Q,K,V dimension. However, as your suggestion, let us build a general search space first, so I replace the `AttentionSuper` with Pytorch's official `MultiHeadAttention`.

File: examples/nas/hub/searchspace/autoformer.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        self.super_mlp_ratio = mlp_ratio
+        self.super_layer_num = depth
+        self.super_num_heads = num_heads
+        self.super_dropout = drop_rate
+        self.super_attn_dropout = attn_drop_rate
+        self.num_classes = num_classes
+        self.pre_norm=pre_norm
+        self.scale=scale
+
+        sample_embed_dim = nn.ValueChoice(search_embed_dim)
```
Comment (matluster) [inline examples/nas/hub/searchspace/autoformer.py:169]: Remove `sample_` to avoid confusion.

File: examples/nas/hub/searchspace/autoformer.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        assert before ^ after
+        if after ^ self.normalize_before:
+            return layer_norm(x)
+        else:
+            return x
+
+
+@model_wrapper
+class AutoformerSpace(nn.Module):
+    def __init__(self, img_size=224, patch_size=16, in_chans=3, num_classes=1000, embed_dim=256, depth=14,
```
Comment (matluster) [inline examples/nas/hub/searchspace/autoformer.py:152]: Suggest adding type-hints for all arguments.

File: examples/nas/hub/searchspace/autoformer.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
+import torch
```
Comment (matluster) [inline examples/nas/hub/searchspace/autoformer.py:1]: I'll test this when the features it relies on are merged, including #4609, #4598, #4524.

If you are in a hurry to merge this in and close this work item, I can talk to @QuanluZhang about merging it first and doing the tests later. Otherwise, let's hold on a minute.

**Comment by nbl97** [comment]
Thanks for your suggestions. I have made the following modifications:

- Re-correct the variable name to avoid confusion, such as removing prefix `sample_` and `super_`.
- Adding type-hints for all arguments.
- Adding some comments for AutoformerSpace.

Please let me know if I can be of any further help.

**Comment by ultmaster** [comment]
We will have a discussion on #4524 very soon. Get back to you after then.

File: nni/retiarii/hub/pytorch/autoformer.py [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+        self.rpe = rpe
+        if rpe:
+            self.rel_pos_embed_k = RelativePosition2D(
+                fixed_embed_dim // num_heads, rpe_length)
+            self.rel_pos_embed_v = RelativePosition2D(
+                fixed_embed_dim // num_heads, rpe_length)
+
+    def forward(self, x):
+        B, N, C = x.shape
+        qkv = self.qkv(x).reshape(
```
Comment (matluster) [inline nni/retiarii/hub/pytorch/autoformer.py:195]: emm. I suggest putting them in the same line.

File: nni/retiarii/hub/pytorch/autoformer.py [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+            return layer_norm(x)
+        else:
+            return x
+
+
+@model_wrapper
+class AutoformerSpace(nn.Module):
+    """
+    The search space that is proposed in Autoformer.
+    There are four searchable variables: depth, embedding dimension, heads number and MLP ratio.
```
Comment (matluster) [inline nni/retiarii/hub/pytorch/autoformer.py:325]: Add a link to autoformer paper.

File: nni/retiarii/hub/pytorch/autoformer.py [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+    search_embed_dim : list of int
+        The search space of embedding dimension.
+    search_mlp_ratio : list of float
+        The search space of MLP ratio.
+    search_num_heads : list of int
+        The search space of number of heads.
+    search_depth: list of int
+        The search space of depth.
+
+    References
```
Comment (matluster) [inline nni/retiarii/hub/pytorch/autoformer.py:364]: We don't use `References` any more.

File: nni/retiarii/hub/pytorch/autoformer.py [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+            # the relative position embedding of V (N, N, head_dim) get shape like (N, B*num_heads, head_dim). We reshape it to the
+            # same size as x (B, num_heads, N, hidden_dim)
+            x = x + (attn_1 @ r_p_v).transpose(1, 0).reshape(B,
+                                                             self.num_heads, N, -1).transpose(2, 1).reshape(B, N, -1)
+        x = self.proj(x)
+        x = self.proj_drop(x)
+        return x
+
+
+class TransformerEncoderLayer(nn.Module):
```
Comment (matluster) [inline nni/retiarii/hub/pytorch/autoformer.py:240]: Suggest adding docstrings here to explain what's the difference between this encoder layer and pytorch's built-in encoder layer.

File: nni/retiarii/hub/pytorch/autoformer.py [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+            device=self.embeddings_table_v.device)
+        # get the embeddings with the corresponding distance
+        embeddings = self.embeddings_table_v[final_mat_v] + \
+            self.embeddings_table_h[final_mat_h]
+
+        return embeddings
+
+
+class Attention(nn.Module):
+    def __init__(
```
Comment (matluster) [inline nni/retiarii/hub/pytorch/autoformer.py:167]: Adding a reference of this "Attention" here.
I suggest naming it "RelativePositionAttention".

File: nni/retiarii/hub/pytorch/autoformer.py [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+        Whether to use pre_norm. Otherwise post_norm is used.
+    global_pool : bool
+        Whether to use global pooling to generate the image representation. Otherwise the cls_token is used.
+    abs_pos : bool
+        Whether to use absolute positional embeddings.
+    qk_scale : float
+        The scaler on score map in self-attention.
+    rpe : bool
+        Whether to use relative position encoding.
+    search_embed_dim : list of int
```
Comment (matluster) [inline nni/retiarii/hub/pytorch/autoformer.py:355]: Please put the searchable dimensions up to the front of the parameter list.

**Comment by QuanluZhang** [comment]
@nbl97 thanks for your contribution! could you resolve the comments?

**Review by JiahangXu — APPROVED** [review]

**Comment by nbl97** [comment]
Hi @QuanluZhang, I have resolved all the comments, pls check it. Thanks~

File: dependencies/recommended.txt [hunk 10]
```diff
@@ -16,3 +16,4 @@ graphviz
 gym
 tianshou >= 0.4.1
 matplotlib
+timm==0.5.4
```
Comment (ultmaster) [inline dependencies/recommended.txt:19]: Avoid ==, use >=. And please add spaces around >=.

File: nni/retiarii/hub/pytorch/autoformer.py [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+        # get the embeddings with the corresponding distance
+        embeddings = self.embeddings_table_v[final_mat_v] + \
+            self.embeddings_table_h[final_mat_h]
+
+        return embeddings
+
+
+class RelativePositionAttention(nn.Module):
+    """
+    This class is designed to support the relative porision in attention.
```
Comment (ultmaster) [inline nni/retiarii/hub/pytorch/autoformer.py:75]: position

File: nni/retiarii/hub/pytorch/autoformer.py [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+    The pytorch built-in nn.MultiheadAttention() does not support relative position embedding.
```
Comment (ultmaster) [inline nni/retiarii/hub/pytorch/autoformer.py:76]: Please brief explain what is RelativePosition

**Comment by nbl97** [comment]
Hi @ultmaster, I have resolved the comments. Thanks for your review~

**Review by ultmaster — APPROVED** [review]
LGTM. Thanks for your contribution!
